aten/src/ATen/cuda/CUDASparseDescriptors.cpp

```diff
@@ -0,0 +1,168 @@
+  auto ndim = input.dim();
+  TORCH_INTERNAL_ASSERT(ndim >= 2);
+  auto rows = input_sizes[ndim - 2];
+  auto cols = input_sizes[ndim - 1];
+
+  bool is_column_major =
+      at::native::is_blas_compatible_column_major_order(input);
+  bool is_row_major = at::native::is_blas_compatible_row_major_order(input);
+  TORCH_INTERNAL_ASSERT(
+      is_column_major || is_row_major,
+      "Expected either row or column major contiguous input.");
+
+  auto leading_dimension =
+      is_row_major ? input_strides[ndim - 2] : input_strides[ndim - 1];
+  auto order = is_row_major ? CUSPARSE_ORDER_ROW : CUSPARSE_ORDER_COL;
+
+  void* values_ptr = input.data_ptr();
+
+  cudaDataType value_type = ScalarTypeToCudaDataType(input.scalar_type());
+  check_supported_cuda_type(value_type);
+
+  cusparseDnMatDescr_t raw_descriptor;
+  TORCH_CUDASPARSE_CHECK(cusparseCreateDnMat(
+      &raw_descriptor,
+      rows,
+      cols,
+      leading_dimension,
+      values_ptr,
+      value_type,
+      order));
+
+  if (ndim > 2) {
+    auto batch_count =
+        at::native::cuda_int_cast(at::native::batchCount(input), "batch_count");
+    auto batch_stride = input_strides[ndim - 3];
+    TORCH_CUDASPARSE_CHECK(cusparseDnMatSetStridedBatch(
+        raw_descriptor, batch_count, batch_stride));
+  }
+
+  descriptor_.reset(raw_descriptor);
+}
+
+CuSparseSpMatCsrDescriptor::CuSparseSpMatCsrDescriptor(const Tensor& input) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.is_sparse_csr());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
+
+  IntArrayRef input_sizes = input.sizes();
+  auto ndim = input.dim();
+  auto rows = input_sizes[ndim - 2];
+  auto cols = input_sizes[ndim - 1];
+  auto nnz = input._nnz();
+
+  auto crow_indices = input.crow_indices();
+  auto col_indices = input.col_indices();
+  auto values = input.values();
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(crow_indices.is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(col_indices.is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(values.is_contiguous());
+
+  cusparseIndexType_t index_type =
+      getCuSparseIndexType(crow_indices.scalar_type());
+  cudaDataType value_type = ScalarTypeToCudaDataType(input.scalar_type());
+  check_supported_cuda_type(value_type);
+
+  cusparseSpMatDescr_t raw_descriptor;
+  TORCH_CUDASPARSE_CHECK(cusparseCreateCsr(
+      &raw_descriptor, // output descriptor
+      rows,
+      cols,
+      nnz,
+      crow_indices
+          .data_ptr(), // row offsets of the sparse matrix, size = rows + 1
+      col_indices.data_ptr(), // column indices of the sparse matrix, size = nnz
+      values.data_ptr(), // values of the sparse matrix, size = nnz
+      index_type, // data type of row offsets index
+      index_type, // data type of col indices
+      CUSPARSE_INDEX_BASE_ZERO, // base index of row offset and col indes
+      value_type // data type of values
+      ));
+
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+  if (ndim > 2) {
```

Comment:
CSR can't have dimension larger than 2 (for now), why do we have this check?

Reply:
Forgot to remove it.